### PR TITLE
CDAP-13296 fix race in program status call

### PR DESCRIPTION
**Pull request by albertshau**
The call to get a program status was suspectible to races because
run records for the program were not read in a single transaction.
If run state happened to change in between run record reads, the
state could incorrectly be returned as stopped.

This race condition would cascade and cause other tests to fail
that would wait for a program run to complete. If one of those
tests ran into this scenario, the test would move on assuming
the program had stopped when it in fact was still running.

Fixed to read run records in a single transaction, and added
missing unit tests.

### CODE REVIEW

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-DUT6463

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    record = RunRecordMeta.builder(record).setStatus(ProgramRunStatus.STARTING).build();
+    status = ProgramLifecycleService.getProgramStatus(Collections.singleton(record));
+    Assert.assertEquals(ProgramStatus.STARTING, status);
+
+    // running -> running
+    record = RunRecordMeta.builder(record).setStatus(ProgramRunStatus.RUNNING).build();
+    status = ProgramLifecycleService.getProgramStatus(Collections.singleton(record));
+    Assert.assertEquals(ProgramStatus.RUNNING, status);
+
+    // failed, killed, completed -> stopped
```
Comment (ecfm) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java]: since the method is called like `getProgramStatus(store.getActiveRuns(programId).values())`, there shouldn't be any record with failed, killed, completed statuses. Better to comment that these cases shouldn't happen but just testing an edge case. The only possible status to cause this method to return `STOPPED` is `SUSPENDED`

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java [hunk 2]
```diff
@@ -479,6 +481,140 @@ public void testgetRuns() throws Exception {
...
+                            actual.containsKey(activeRun.getKey().getParent()));
+          actual.get(programId).add(activeRun.getValue().getStatus());
+        }
+
+        Assert.assertEquals(expected, actual);
+      });
+    }
+
+    // check active runs per app
+    for (ApplicationId app : Arrays.asList(ns1.app(app1), ns1.app(app2), ns2.app(app1), ns2.app(app2))) {
```
Comment (ecfm) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java]: can get the appId's from transforming the programId's into appId's and collect them as a set

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java (hunk 1 above)
Comment (albertshau) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java]: I wouldn't call that an edge case, as it's the point of the test, to check that it returns the correct things

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java (hunk 2 above)
Comment (albertshau) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java]: is the concern if we add program ids but don't modify this list? I can create an app id collection first then create the program id collection from that.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java (hunk 1 above)
Comment (ecfm) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java]: right, I mean better to clarify these statuses shouldn't occur in real usage, and add a test case for suspended -> stopped

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java [hunk 3]
```diff
@@ -479,6 +484,141 @@ public void testgetRuns() throws Exception {
...
+          store.recordProgramProvisioning(runId, Collections.emptyMap(), Collections.emptyMap(),
+                                          AppFabricTestHelper.createSourceId(sourceId.incrementAndGet()), ARTIFACT_ID);
+          store.recordProgramProvisioned(runId, 3, AppFabricTestHelper.createSourceId(sourceId.incrementAndGet()));
+          twillRunId = UUID.randomUUID().toString();
+          store.recordProgramStart(runId, twillRunId, Collections.emptyMap(),
+                                   AppFabricTestHelper.createSourceId(sourceId.incrementAndGet()));
+          store.recordProgramStop(runId, System.currentTimeMillis(), runStatus, null,
+                                  AppFabricTestHelper.createSourceId(sourceId.incrementAndGet()));
+        }
+      });
```
Comment (ecfm) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java:560]: better to add a suspended record here

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java (hunk 1 above)
Comment (albertshau) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java]: I don't think suspend is actually implemented for anything. The old logic certainly did not handle those run statuses. I can add that, but suspended and resuming should not be stopped, they should be running.

I also disagree about adding a comment that this method normally gets only active run records, as the method does not and should not care what type of run records it gets. Instead, I can rename 'activeRunRecords' to just 'runRecords', since that is more clear.

Comment (albertshau) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/ProgramLifecycleServiceTest.java]: Ok a workflow can be suspended, which wasn't immediately clear to me. That means the previous logic was buggy, as it would say a suspended or resuming program was stopped, and disallow stopping that run.

**Comment by ecfm** [comment]
LGTM

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java (hunk 3 above)
Comment (albertshau) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java:560]: done
